### PR TITLE
Proper render pass support

### PR DESCRIPTION
**Pull request by kvark**
This PR implements the API scheme of https://github.com/gpuweb/gpuweb/pull/102 and caches both render passes and framebuffers properly in the device.

### CODE REVIEW

File: wgpu-native/src/device.rs [hunk 1]
```diff
@@ -14,9 +14,17 @@ use hal::{self, Device as _Device};
...
 use std::{ffi, slice};
+use std::collections::hash_map::{Entry, HashMap};
 use std::sync::Mutex;
 
 
+#[derive(Hash, PartialEq)]
+pub(crate) struct RenderPassKey {
+    pub attachments: Vec<hal::pass::Attachment>,
+}
+impl Eq for RenderPassKey {}
```
Comment (grovesNL) [inline wgpu-native/src/device.rs]: nit: newline

File: wgpu-native/src/device.rs [hunk 2]
```diff
@@ -464,11 +421,78 @@ pub extern "C" fn wgpu_device_create_render_pipeline(
...
+            attachments: color_keys.chain(depth_stencil_key).collect(),
+        }
+    };
+
+    let mut render_pass_cache = device.render_passes.lock().unwrap();
+    let main_pass = match render_pass_cache.entry(rp_key) {
+        Entry::Occupied(e) => e.into_mut(),
+        Entry::Vacant(e) => {
+            let color_ids = [
+                (0, hal::image::Layout::ColorAttachmentOptimal),
```
Comment (grovesNL) [inline wgpu-native/src/device.rs]: Why hardcode 4 here? Or is it just temporary?

File: wgpu-native/src/device.rs [hunk 3]
```diff
@@ -24,6 +24,12 @@ pub(crate) struct RenderPassKey {
 }
 impl Eq for RenderPassKey {}
 
+#[derive(Hash, PartialEq)]
+pub(crate) struct FramebufferKey {
+    pub attachments: Vec<Stored<TextureViewId>>,
+}
+impl Eq for FramebufferKey {}
```
Comment (grovesNL) [inline wgpu-native/src/device.rs]: nit: newline

**Comment by kvark** [comment]
Yeah, 4 is a quick temporary.
Did you finish the review or still WIP?

**Comment by kvark** [comment]
Thanks!
Bors r=grovesNL

**Pull request merged**